pkg/scheduler/framework/v1alpha1/interface.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file defines the scheduling framework plugin interfaces.
+
+package v1alpha1
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	internalcache "k8s.io/kubernetes/pkg/scheduler/internal/cache"
+)
+
+// Plugin is the parent type for all the scheduling framework plugins.
+type Plugin interface {
+	Name() string
+}
+
+// ReservePlugin is an interface for Reserve plugins. These plugins are called
+// at the reservation point, AKA "assume". These are meant to updated the state
+// of the plugin. They do not return any value (other than error).
+type ReservePlugin interface {
+	Plugin
+	// Reserve is called by the scheduling framework when the scheduler cache is
+	// updated.
+	Reserve(pc *PluginContext, p *v1.Pod, nodeName string) error
+}
+
+// PrebindPlugin is an interface that must be implemented by "prebind" plugins.
+// These plugins are called before a pod being scheduled
+type PrebindPlugin interface {
+	Plugin
+	// Prebind is called before binding a pod. All prebind plugins must return
+	// failure = false, err = nil or the pod will be rejected and won't be sent
+	// for binding. "failure" indicates that an unexpected failure has happened.
+	// Plugins that run successfully and want to reject a pod should return
+	// false, non-nil error.
+	Prebind(pc *PluginContext, p *v1.Pod, nodeName string) (failure bool, err error)
+}
+
+// SchedulerCache is an interface that provides read-only access to the scheduler
+// cache for plugins.
+// NOTE: Scheduler cache contains live data. The cache is snapshotted at the
+// beginning of a scheduling cycle and then the contents of that snapshot is
+// passed to plugins when plugins are called by the framework. If plugins use this
+// interface to read the cache directly, they may receive some information which
+// is different from what is passed to them by the framework.
+type SchedulerCache interface {
```

Comment:
I've been trying to figure out why I don't like this interface name, and I think I've had a realization.

### "Framework" means more than one thing **within the scheduling framework**!

From the perspective of the scheduler implementation, "framework" is the design/architecture that lets us delegate feature implementations to plugins. From the perspective of a plugin, "framework" is the set of APIs that lets a plugin do plugin things. From the outside, "framework" means the combination of the previous two definitions.

To be honest, I'm surprised the term "SDK" hasn't crept into this feature yet. 

Arguing over names may seem trivial, but I believe in cases like this it will have a significant impact on how maintainable the scheduling framework turns out to be.

Sorry this comment doesn't include suggested names, but it's late and I haven't had dinner.

Reply:
Right now we are giving plugins access to `ScheduleCache` and `NodeInfoSnapshot`. "Cache" and "snapshot" are similar concepts, and the differences are not clear from their names. As we discussed, I think the two concepts should be combined into one interface, but that presents challenges about the freshness or consistency of the data.